src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -385,38 +384,46 @@ void DatabaseCatalog::updateUUIDMapping(const UUID & uuid, DatabasePtr database,
     it->second = std::make_pair(std::move(database), std::move(table));
 }
 
+std::unique_ptr<DatabaseCatalog> DatabaseCatalog::database_catalog;
+
 DatabaseCatalog::DatabaseCatalog(Context * global_context_)
     : global_context(global_context_), log(&Poco::Logger::get("DatabaseCatalog"))
 {
     if (!global_context)
-        throw Exception("DatabaseCatalog is not initialized. It's a bug.", ErrorCodes::NULL_POINTER_DEREFERENCE);
+        throw Exception("DatabaseCatalog is not initialized. It's a bug.", ErrorCodes::LOGICAL_ERROR);
 }
 
 DatabaseCatalog & DatabaseCatalog::init(Context * global_context_)
 {
-    static DatabaseCatalog database_catalog(global_context_);
-    return database_catalog;
+    if (database_catalog)
```

Comment:
You don't try to avoid race if there is a bug?

Reply:
Race between what and what?

Reply:
If you call init() twice in parallel, you can check is is not initialized and create DatabaseCatalog from different threads.

Reply:
Yep, that's what I'm asking -- when is init called twice in parallel, and from which threads?

Reply:
It is not thread-safe, but `DatabaseCatalog::init()` is called by main thread only once on early stages of startup, so thread-safe singleton initialization actually is not required here.